### PR TITLE
feat: Maintain tagging in methods inherited from `Table` class

### LINKED ISSUE #58: In `TaggedTable` override methods of `Table` that return a `Table`
### Is your feature request related to a problem?

Now that `TaggedTable` is a subclass of `Table`, it also inherits some methods that return a new `Table`. However, currently those methods lose the tags, since a `Table` simply doesn't know about them. Methods that remove tagged `Column`s must 
throw an `IllegalSchemaModificationException`.

### Desired solution

Override the methods and return a new `TaggedTable` instead.

### Possible alternatives (optional)

_No response_

### Screenshots (optional)

_No response_

### Additional Context (optional)

_No response_

### PR DESCRIPTION
**Pull request by zzril**
Closes #58.

### Summary of Changes

* Overrode methods in `TaggedTable` inherited from `Table` to maintain tagging.
* Introduced new exceptions to be thrown when the overriden method cannot be applied
* Added tests.

### CODE REVIEW

**Comment by lars-reimann** [comment]
## [🦙 MegaLinter](https://megalinter.io/6.22.2) status: ✅ [SUCCESS](https://github.com/Safe-DS/Stdlib/actions/runs/5111343330)

| Descriptor  |                                 Linter                                 |Files|Fixed|Errors|Elapsed time|
|-------------|------------------------------------------------------------------------|-----|----:|------|------------|
|✅ PYTHON    |[black](https://megalinter.io/6.22.2/descriptors/python_black)          |    2|    0|     0|0.57s       |
|✅ PYTHON    |[mypy](https://megalinter.io/6.22.2/descriptors/python_mypy)            |    2|     |     0|1.38s       |
|✅ PYTHON    |[ruff](https://megalinter.io/6.22.2/descriptors/python_ruff)            |    2|    1|     0|0.04s       |
|✅ REPOSITORY|[git_diff](https://megalinter.io/6.22.2/descriptors/repository_git_diff)|yes  |     |no    |0.03s       |

See detailed report in [MegaLinter reports](https://github.com/Safe-DS/Stdlib/actions/runs/5111343330)
_Set `VALIDATE_ALL_CODEBASE: true` in mega-linter.yml to validate all sources, not only the diff_

_MegaLinter is graciously provided by [OX Security](https://www.ox.security/?ref=megalinter)_

File: tests/safeds/data/tabular/containers/_table/_tagged_table/test_filter_rows.py
Comment (lars-reimann) [inline tests/safeds/data/tabular/containers/_table/_tagged_table/test_filter_rows.py:27]: Might be worth implementing `__eq__` for `TaggedTable` first so only one assertion is needed.

Comment (zzril) [reply tests/safeds/data/tabular/containers/_table/_tagged_table/test_filter_rows.py:27]: I thought we had deliberately decided against that, so failing tests would produce more helpful output.

Wasn't this the point of #124?

**Comment by zzril** [comment]
Note:  
Some of the methods that can remove a column (`keep_only_columns`, `remove_columns`) are a bit more intricate as they seem to require adjusting lots of existing tests for the `Table` class.  
May want to wait with this until #314 is merged.

File: tests/safeds/data/tabular/containers/_table/_tagged_table/test_filter_rows.py
Comment (lars-reimann) [reply tests/safeds/data/tabular/containers/_table/_tagged_table/test_filter_rows.py:27]: Works for me, too.

Comment (zzril) [reply tests/safeds/data/tabular/containers/_table/_tagged_table/test_filter_rows.py:27]: Thinking more about this, it could make sense to extract it into its own helper method in `tests/helpers`, though.

**Comment by zzril** [comment]
Note: The overriden version of `transform_table` in the `TaggedTable` class will throw an exception if the transformer (e.g. `OneHotEncoder`) removed the target column.  
The fact that this will happen can already be detected in the `fit`ting phase. It may make sense to test for this case in he `Transformer`'s `fit` method and raise a warning.

File: tests/safeds/data/tabular/containers/_table/_tagged_table/test_filter_rows.py
Comment (zzril) [reply tests/safeds/data/tabular/containers/_table/_tagged_table/test_filter_rows.py]: Did extract it into its own helper method.

**Comment by zzril** [comment]
The subclass `TaggedTable` throwing additional exceptions seems like a bad idea (see [this comment](https://github.com/Safe-DS/Stdlib/issues/58#issuecomment-1572727668)).

I'm currently implementing an alternative solution where methods such as `remove_columns` are not overriden, but we introduce new methods such as `remove_feature_columns` that would maintain the tagging, on [this branch](https://github.com/Safe-DS/Stdlib/tree/58b-don-t-override-when-throwing).

Merge that branch (58b) into this one (58) once it's clear we want to implement it that way.

**Comment by zzril** [comment]
### Notes about the last merge:

A new check was added on `main` to the `test_tagged_table.py` file that we deleted.
Need to move that to an appropriate place in the `tagged_table` subfolder.

May want to rethink current orderung of functions in `_tagged_table.py`. The `copy` method may just belong to the group of overriden methods.

We also need to manually merge any updates on the docstrings from `_table.py` into ours. `git` cannot mark these as conflicting because our functions do not yet exist in `main`, but we ofc want our dosctrings s to be consistent with those form `table.py`.

File: src/safeds/data/tabular/containers/_tagged_table.py
Comment (patrikguempel) [inline src/safeds/data/tabular/containers/_tagged_table.py:742]:
```suggestion
        * If the function returns a negative number, `col1` will be ordered before `col2`.
        * If the function returns a positive number, `col1` will be ordered after `col2`.
        * If the function returns 0, the original order of `col1` and `col2` will be kept.
```

File: src/safeds/data/tabular/containers/_tagged_table.py
Comment (patrikguempel) [inline src/safeds/data/tabular/containers/_tagged_table.py:777]:
```suggestion
        * If the function returns a negative number, `row1` will be ordered before `row2`.
        * If the function returns a positive number, `row1` will be ordered after `row2`.
        * If the function returns 0, the original order of `row1` and `row2` will be kept.
```

File: src/safeds/exceptions/_data.py
Comment (patrikguempel) [inline src/safeds/exceptions/_data.py:144]:
```suggestion
    """Exception raised when modifying a schema in a way that is inconsistent with the subclass's requirements."""
```

File: src/safeds/data/tabular/containers/_tagged_table.py
Comment (Marsmaennchen221) [reply src/safeds/data/tabular/containers/_tagged_table.py:497]:
```suggestion
        Remove every column that contains non-numerical values.
```

File: src/safeds/data/tabular/containers/_tagged_table.py
Comment (Marsmaennchen221) [reply src/safeds/data/tabular/containers/_tagged_table.py:529]:
```suggestion
        Remove all row duplicates.
```

**Review by patrikguempel — APPROVED** [review]
😊

File: src/safeds/data/tabular/containers/_tagged_table.py
Comment (sibre28) [inline src/safeds/data/tabular/containers/_tagged_table.py:420]:
```suggestion
        return TaggedTable._from_table(
            super().keep_only_columns(column_names),
            target_name=self.target.name,
            feature_names=sorted(
                set(self.features.column_names).intersection(set(column_names)),
```

File: src/safeds/data/tabular/containers/_tagged_table.py
Comment (sibre28) [inline src/safeds/data/tabular/containers/_tagged_table.py:451]:
```suggestion
            raise IllegalSchemaModificationError("You cannot remove the target column.")
```

File: src/safeds/data/tabular/containers/_tagged_table.py
Comment (sibre28) [inline src/safeds/data/tabular/containers/_tagged_table.py:625]:
```suggestion
        If the column to be replaced is the target column, it must be replaced by exactly one column. This column becomes the new target column.
```

File: src/safeds/data/tabular/containers/_tagged_table.py
Comment (sibre28) [inline src/safeds/data/tabular/containers/_tagged_table.py:626]:
```suggestion
If the column to be replaced is a feature column, the new columns that replace it all become feature columns.
```

File: tests/safeds/data/tabular/containers/_table/_tagged_table/test_features.py
Comment (sibre28) [inline tests/safeds/data/tabular/containers/_table/_tagged_table/test_features.py:34]:
```suggestion
    ids=["all_columns_except_target_are_features", "not_all_columns_are_features"],
```

File: tests/safeds/data/tabular/containers/_table/_tagged_table/test_keep_only_columns.py
Comment (sibre28) [inline tests/safeds/data/tabular/containers/_table/_tagged_table/test_keep_only_columns.py:81]:
```suggestion
    ids=["keep_feature_and_target_column", "keep_non_feature_column", "don't_keep_non_feature_column"],
```

File: tests/safeds/data/tabular/containers/_table/_tagged_table/test_remove_columns_with_non_numerical_values.py
Comment (sibre28) [inline tests/safeds/data/tabular/containers/_table/_tagged_table/test_remove_columns_with_non_numerical_values.py:169]:
```suggestion
        "all_features_non_numerical",
        "all_features_and_non_feature_non_numerical",
```

File: src/safeds/data/tabular/containers/_tagged_table.py
Comment (sibre28) [inline src/safeds/data/tabular/containers/_tagged_table.py:412]: Consider raising a ColumnIsTargetError here instead

Comment (zzril) [reply src/safeds/data/tabular/containers/_tagged_table.py:412]: I think I had decided against this because the `ColumnIsTargetError` explicitely talks about "removing" a column.
As we're not actively removing columns (that's rather a side effect of us *selecting* and keeping a subset of the columns), I thought the error message format of the `ColumnIsTargetError` doesn't fit 100% and might be confusing.
With the general `IllegalSchemaModificationError`, I have more liberties concerning the exact wording.

I'll let you decide.

File: tests/safeds/data/tabular/containers/_table/_tagged_table/test_features.py
Comment (zzril) [reply tests/safeds/data/tabular/containers/_table/_tagged_table/test_features.py]: Solved differently.

File: src/safeds/data/tabular/containers/_tagged_table.py
Comment (sibre28) [reply src/safeds/data/tabular/containers/_tagged_table.py:412]: I see, makes sense to me!

File: tests/safeds/data/tabular/containers/_table/_tagged_table/test_features.py
Comment (sibre28) [reply tests/safeds/data/tabular/containers/_table/_tagged_table/test_features.py]: k

**Review by sibre28 — APPROVED** [review]
[LGTM](https://lgtm.party/p/SAT66D9/)

**Pull request merged**
